### PR TITLE
[1/2] Ray trainer

### PR DESCRIPTION
**Pull request by coledie**
- [x] Forked repo is synced with upstream -> github shows no code delta outside of the desired.
    https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork
- [x] Tests are passing? https://github.com/microsoft/DeepGNN/blob/main/CONTRIBUTING.md#run-tests
- [x] Changelog and documentation updated.
- [x] PR is labeled using the label menu on the right side.

Previous Behavior
----------------
We use custom built trainers.

New Behavior
----------------
* Use ray train as trainers https://docs.ray.io/en/latest/train/getting-started.html
* Refactor BaseModel to be more inline with common usage.

### CODE REVIEW

File: src/python/deepgnn/pytorch/modeling/base_model.py [hunk 1]
```diff
@@ -121,9 +117,7 @@ def encode_feature(self, context: dict):
         if self.feature_enc:
             self.feature_enc.forward(context)
 
-    def forward(
-        self, context: QueryOutput
-    ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+    def forward(self, context: QueryOutput) -> torch.Tensor:
```
Comment (mortid0) [inline src/python/deepgnn/pytorch/modeling/base_model.py:120]: That's actually a big change, because I assumed all models inherit from base model. Should we just remove it and rely on `nn.Module.forward`?

File: docs/torch/node_class.rst [hunk 2]
```diff
@@ -34,261 +44,221 @@ DeepGNN models typically contain multiple parts:
...
-	...         drop_last=True,
-	...         worker_index=rank,
-	...         num_workers=world_size,
-	...     )
-
-Arguments
-=========
-`init_args` registers any model specific arguments.
+Here we define our training function.
+In the setup part we do 2 notable things things,
```
Comment (mortid0) [inline docs/torch/node_class.rst:188]: nit: 2->two

File: docs/torch/node_class.rst [hunk 3]
```diff
@@ -34,261 +44,221 @@ DeepGNN models typically contain multiple parts:
...
-	>>> def init_args(parser):
-	...     parser.add_argument("--head_num", type=str2list_int, default="8,1", help="the number of attention headers.")
-	...     parser.add_argument("--hidden_dim", type=int, default=8, help="hidden layer dimension.")
-	...     parser.add_argument("--num_classes", type=int, default=-1, help="number of classes for category")
-	...     parser.add_argument("--ffd_drop", type=float, default=0.0, help="feature dropout rate.")
-	...     parser.add_argument("--attn_drop", type=float, default=0.0, help="attention layer dropout rate.")
-	...     parser.add_argument("--l2_coef", type=float, default=0.0005, help="l2 loss")
-	...     parser.add_argument("--neighbor_edge_types", type=str2list_int, default="0", help="Graph Edge for attention encoder.",)
-	...     parser.add_argument("--eval_file", default="", type=str, help="")
+* Initialize the ray dataset, see more details in `docs/graph_engine/dataset.rst`.
```
Comment (mortid0) [inline docs/torch/node_class.rst:192]: `docs/graph_engine/dataset.rst` is a graph engine dataset, not ray. Wrong link?

File: docs/torch/node_class.rst [hunk 4]
```diff
@@ -2,11 +2,15 @@
...
 
 Cora Dataset
 ============
-The Cora dataset consists of 2708 scientific publications represented as nodes interconnected by 5429 reference links represented as edges. Each paper is described by a binary mask for 1433 pertinent dictionary words and an integer in {0..6} representing its type.
+The Cora dataset consists of 2708 scientific publications represented as nodes interconnected by 5429
+reference links represented as edges. Each paper is described by a binary mask for 1433 pertinent
+dictionary words and an integer in {0..6} representing its type.
 First we download the Cora dataset and convert it to a valid binary representation via our built-in Cora downloader.
 
 .. code-block:: python
```
Comment (mortid0) [inline docs/torch/node_class.rst:16]: We don't need this one:
```python
>>> Cora("/tmp/cora/")
```
looks like you are using in memory clients below instead of a distributed setup.

File: docs/torch/node_class.rst [hunk 5]
```diff
@@ -34,261 +44,221 @@ DeepGNN models typically contain multiple parts:
...
+    ...     def forward(self, context: Dict[Any, np.ndarray]):
+    ...         nodes = torch.squeeze(context["nodes"])                # [N], N: num of nodes in subgraph
+    ...         feat = torch.squeeze(context["feat"])                  # [N, F]
+    ...         mask = torch.squeeze(context["input_mask"])            # [N]
+    ...         labels = torch.squeeze(context["labels"])              # [N]
+    ...         edges = torch.squeeze(context["edges"].reshape((-1, 2)))                # [X, 2], X: num of edges in subgraph
+    ...
+    ...         edges = np.transpose(edges)
+    ...
+    ...         # TODO This is not stable, when doing batch_size < graph size ends up with size < index values. use torch.unique to remap edges
```
Comment (mortid0) [inline docs/torch/node_class.rst:175]: todo in this PR?

File: docs/torch/node_class.rst [hunk 6]
```diff
@@ -34,261 +44,221 @@ DeepGNN models typically contain multiple parts:
...
-	... )
+    >>> ray.init()
+    RayContext(...)
+    >>> trainer = TorchTrainer(
+    ...     train_func,
+    ...     train_loop_config={},
+    ...     run_config=RunConfig(verbose=0),
+    ...     scaling_config=ScalingConfig(num_workers=1, use_gpu=False),
+    ... )
+    >>> result = trainer.fit()
```
Comment (mortid0) [inline docs/torch/node_class.rst:264]: what is the expected outcome? Are we going to evaluate our model and check accuracy?

File: docs/torch/node_class.rst [hunk 7]
```diff
@@ -34,261 +44,221 @@ DeepGNN models typically contain multiple parts:
...
-	...     init_args_fn=init_args,
-	...		run_args=arg_list,
-	... )
+    >>> ray.init()
+    RayContext(...)
+    >>> trainer = TorchTrainer(
+    ...     train_func,
+    ...     train_loop_config={},
+    ...     run_config=RunConfig(verbose=0),
+    ...     scaling_config=ScalingConfig(num_workers=1, use_gpu=False),
```
Comment (mortid0) [inline docs/torch/node_class.rst:262]: Can we use 2 workers for training?

File: docs/torch/node_class.rst (hunk 5 above)
Comment (coledie) [reply docs/torch/node_class.rst:175]: It's an issue with GAT model not this example specifically, removed.

File: docs/torch/node_class.rst (hunk 4 above)
Comment (coledie) [reply docs/torch/node_class.rst:16]: This is to generatte the Cora dataset so we can use it in the example.

Comment (mortid0) [reply docs/torch/node_class.rst:16]: We generate it anyway later, right?

**Review by mortid0 — COMMENTED** [review]
Could you reply to comments/get CI passing before requesting review?

File: docs/torch/node_class.rst (hunk 4 above)
Comment (coledie) [reply docs/torch/node_class.rst:16]: We use distro client now, in next PR.

File: docs/torch/node_class.rst (hunk 6 above)
Comment (coledie) [reply docs/torch/node_class.rst:263]: I've added eval step.

File: docs/torch/node_class.rst [hunk 8]
```diff
@@ -2,28 +2,40 @@
 Node Classification with GAT
```
[the remark this answers is not in the document]
Comment (coledie) [reply docs/torch/node_class.rst:2]: The goal of this PR is to replace the old trainers with Ray

[the remark this answers is not in the document]
Comment (mortid0) [reply docs/torch/node_class.rst:2]: Where are changes in old trainers? I don't see relevant updates:
image

[the remark this answers is not in the document]
Comment (coledie) [reply docs/torch/node_class.rst:2]: Final 3 PRs will be to add migrate scripts for ge, torch and tf - then I will delete all the old stuff. If I do it now all the tests will fail.

File: docs/torch/ray_usage.rst [hunk 9]
```diff
@@ -0,0 +1,297 @@
+****************************
+Node Classification with GAT
+****************************
+
+In this guide we use a pre-built `Graph Attention Network(GAT) <https://arxiv.org/abs/1710.10903>`_ model
```
Comment (mortid0) [inline docs/torch/ray_usage.rst:5]: Please, reference the original documentation for node classification and avoid copy-pasting. You can also use some other model to show how to work with ray.

File: docs/torch/ray_usage.rst [hunk 10]
```diff
@@ -0,0 +1,297 @@
+****************************
```
Comment (mortid0) [inline docs/torch/ray_usage.rst:1]: Looks like this document is missing from `docs/torch/index.rst`.

**Review by mortid0 — CHANGES_REQUESTED** [review]

File: docs/torch/ray_usage.rst (hunk 9 above)
Comment (coledie) [reply docs/torch/ray_usage.rst:5]: I'll reference original documentation and update the title, I should keep the GAT model though since we are going to replace old documentation in the near future anyway - it is a lot of work to get a new model in with Ray Data.

File: CHANGELOG.md [hunk 11]
```diff
@@ -7,6 +7,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [Unreleased]
 
 ### Changed
+- Add usage example for Ray Train, see [docs/torch/ray_usage.rst](https://github.com/microsoft/DeepGNN/tree/main/docs/torch/ray_usage.rst).
```
Comment (mortid0) [inline CHANGELOG.md:10]: nit: looks like this PR needs a rebase.

File: docs/torch/ray_usage.rst [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+    ...         return scores
+
+
+Ray Train
+=========
+
+Here we define our training function.
+In the setup part we do two notable things things,
+
+* Wrap the model and optimizer with train.torch.prepare_model/optimizer for Ray multi worker usage.
```
Comment (mortid0) [inline docs/torch/ray_usage.rst:150]: `train.torch.prepare_model/optimizer` - make it a link to ray docs?

File: docs/torch/ray_usage.rst [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+    ...     train_loop_config={
+    ...         "data_dir": data_dir.name,
+    ...         "sample_filename": "test.nodes",
+    ...         "n_epochs": 1,
+    ...         "model_dir": f"{model_dir.name}/model.pt",
+    ...     },
+    ...     run_config=RunConfig(verbose=0),
+    ...     scaling_config=ScalingConfig(num_workers=1, use_gpu=False),
+    ... )
+    >>> result = trainer.fit()
```
[the remark this answers is not in the document]
Comment (coledie) [reply docs/torch/ray_usage.rst:253]: I've added metric + loss logging to the doc.

File: docs/torch/ray_usage.rst [hunk 14]
```diff
@@ -0,0 +1,256 @@
+*****************
+Ray Usage Example
+*****************
+
+In this guide we use a pre-built `Graph Attention Network(GAT) <https://arxiv.org/abs/1710.10903>`_ model
+to classify nodes in the `Cora dataset <https://graphsandnetworks.com/the-cora-dataset/>`_. This is the same
+as our `node classification example <https://github.com/microsoft/DeepGNN/blob/main/docs/torch/node_class.rst>` except here we use Ray as the trainer.
```
[the remark this answers is not in the document]
Comment (mortid0) [reply docs/torch/ray_usage.rst]: image
You are referencing a github page, not readthedocs..

[the remark this answers is not in the document]
Comment (mortid0) [reply docs/torch/ray_usage.rst]: https://docs.readthedocs.io/en/stable/guides/cross-referencing-with-sphinx.html#the-doc-role

**Review by mortid0 — APPROVED** [review]
